src/darwin/Framework/CHIPTests/MTRDeviceTests.m

```diff
@@ -2605,6 +2606,166 @@ - (void)test028_TimeZoneAndDST
+    [self waitForExpectations:[NSArray arrayWithObject:expectation] timeout:kDownloadLogTimeoutInSeconds];
+}
+
+- (void)test030_DownloadNetworkDiagnosticsLog_NoTimeout
+{
+    MTRDeviceController * controller = sController;
+    XCTAssertNotNil(controller);
+    XCTestExpectation * expectation = [self expectationWithDescription:@"Network Diagnostics Log Transfer Complete"];
+
+    dispatch_queue_t queue = dispatch_get_main_queue();
+
+    dispatch_async(queue, ^{
+        MTRDevice * device = [MTRDevice deviceWithNodeID:@(kDeviceId) controller:controller];
+        XCTAssertNotNil(device);
+
+        NSString * outFile = [NSString stringWithFormat:@"/tmp/networkdiagnosticslog.txt"];
+        NSError * error = [self generateLogFile:outFile size:(4 * 1024)];
+
+        if (error != nil) {
+            NSLog(@"Failed to generate log file");
+            return;
+        }
+
+        [device downloadLogOfType:MTRDiagnosticLogTypeNetworkDiagnostics timeout:0 queue:queue completion:^(NSURL * _Nullable logResult, NSError * error) {
+            XCTAssertNil(error);
+            XCTAssertNotNil(logResult);
+
+            NSError * attributesError = nil;
+            NSDictionary * fileAttributes = [[NSFileManager defaultManager] attributesOfItemAtPath:[logResult path] error:&attributesError];
+
+            size_t fileSize = [fileAttributes fileSize];
+            XCTAssertTrue(fileSize > 0);
+            [expectation fulfill];
+        }];
+    });
+    [self waitForExpectations:[NSArray arrayWithObject:expectation] timeout:kDownloadLogTimeoutInSeconds];
+}
+
+- (void)test031_DownloadCrashLog_NoTimeout
+{
+    MTRDeviceController * controller = sController;
+    XCTAssertNotNil(controller);
+    XCTestExpectation * expectation = [self expectationWithDescription:@"Crash Log Transfer Complete"];
+
+    dispatch_queue_t queue = dispatch_get_main_queue();
+
+    dispatch_async(queue, ^{
+        MTRDevice * device = [MTRDevice deviceWithNodeID:@(kDeviceId) controller:controller];
+        XCTAssertNotNil(device);
+
+        NSString * outFile = [NSString stringWithFormat:@"/tmp/crashlog.txt"];
+        NSError * error = [self generateLogFile:outFile size:(3 * 1024)];
+
+        if (error != nil) {
+            NSLog(@"Failed to generate log file");
+            return;
+        }
+
+        [device downloadLogOfType:MTRDiagnosticLogTypeCrash timeout:0 queue:queue completion:^(NSURL * _Nullable logResult, NSError * error) {
+            XCTAssertNil(error);
+            XCTAssertNotNil(logResult);
+
+            NSError * attributesError = nil;
+            NSDictionary * fileAttributes = [[NSFileManager defaultManager] attributesOfItemAtPath:[logResult path] error:&attributesError];
+
+            size_t fileSize = [fileAttributes fileSize];
+            XCTAssertTrue(fileSize > 0);
+            [expectation fulfill];
+        }];
+    });
+    [self waitForExpectations:[NSArray arrayWithObject:expectation] timeout:kDownloadLogTimeoutInSeconds];
+}
+
```

Comment:
Missing tests:

1. Starting two downloads in parallel.
2. Various error cases.

Reply:
Will file issues for them. Starting two downloads using BDX can't be supported with Responder.

Reply:
filed #30543